bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirSerialConnector.java

```diff
@@ -0,0 +1,590 @@
+
+    private static byte CTRL = (byte) 0x07;
+    private static byte[] START = { CTRL, (byte) 0xf0 };
+    private static byte[] END = { CTRL, (byte) 0x0f };
+    private static byte[] ACK = { CTRL, (byte) 0xf3 };
+
+    private static int RS232_ENABLED_VALUE = 0x03;
+    private static int RS232_DISABLED_VALUE = 0x00;
+
+    private boolean isSuspended = true;
+
+    private boolean connected = false;
+    private final String serialPortName;
+    private final int baudRate;
+    private final SerialPortManager serialPortManager;
+    private @Nullable SerialPort serialPort;
+    private @Nullable InputStream inputStream;
+    private @Nullable OutputStream outputStream;
+
+    public ComfoAirSerialConnector(final SerialPortManager serialPortManager, final String serialPortName,
+            final int baudRate) {
+        this.serialPortManager = serialPortManager;
+        this.serialPortName = serialPortName;
+        this.baudRate = baudRate;
+    }
+
+    /**
+     * Open serial port.
+     */
+    public void open() {
+        logger.debug("open(): Opening ComfoAir connection");
+
+        try {
+            SerialPortIdentifier portIdentifier = serialPortManager.getIdentifier(serialPortName);
+            if (portIdentifier != null) {
+                SerialPort serialPort = portIdentifier.open(this.getClass().getName(), 3000);
+                serialPort.setSerialPortParams(baudRate, SerialPort.DATABITS_8, SerialPort.STOPBITS_1,
+                        SerialPort.PARITY_NONE);
+                serialPort.enableReceiveThreshold(1);
+                serialPort.enableReceiveTimeout(1000);
+                serialPort.notifyOnDataAvailable(true);
+                this.serialPort = serialPort;
+
+                inputStream = new DataInputStream(new BufferedInputStream(serialPort.getInputStream()));
+                outputStream = serialPort.getOutputStream();
+
+                ComfoAirCommand command = ComfoAirCommandType.getChangeCommand(ComfoAirCommandType.ACTIVATE.getKey(),
+                        OnOffType.ON);
+
+                if (command != null) {
+                    sendCommand(command, ComfoAirCommandType.Constants.EMPTY_INT_ARRAY);
+                } else {
+                    logger.debug("Failure while creating COMMAND: {}", command);
+                }
+                setConnected(true);
+            } else {
+                logger.debug("open(): No such Port: {}", serialPortName);
+                setConnected(false);
+            }
+        } catch (PortInUseException e) {
+            logger.debug("open(): Port in Use Exception: {}", e.getMessage(), e);
+            setConnected(false);
+        } catch (UnsupportedCommOperationException e) {
+            logger.debug("open(): Unsupported Comm Operation Exception: {}", e.getMessage(), e);
+            setConnected(false);
+        } catch (IOException e) {
+            logger.debug("open(): IO Exception: {}", e.getMessage(), e);
+            setConnected(false);
```

Comment:
It seems a bit unreasonable to log the stack trace e.g. when the serial port is in use.

Suggestion: Throw these exceptions and handle them in `ComfoAirHandler`. You can set the thing offline with the exception's message, which will be displayed in the UI, then.

Reply:
Looking at some other bindings, I'm not sure what would be the best approach here. From my point of view there are three options:

1. Just rethrow each Exception and catch them in the Handler individually
2. Rethrow them all as a more generic Exception, e.g. an IOException
3. Add a new `ComfoAirException` class and rethrow as such

What would you recommend?

Reply:
I'd say it's a matter of taste. My favorite is 1., because I think it's the most simple approach with respect to code lines. I use multi-tray/catches with the |-syntax to handle the different exception types in the handler.